accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,279 @@
+
+Since then, we've seen requests from partners who have tried using it and hit
+limitations that come from guardrails we put into place due to the complex
+nature of platform-level preview features.
+
+The goal of this feature is to find a design that works for both, platform-level
+preview features and library-level preview features.
+
+Let's first define these two terms:
+
+* **platform-level preview feature**. These are features that rely on preview
+  behaviors across runtime, core library, tooling, and languages. The nature of
+  these features is that they are in support of each other. For example, generic
+  math requires new runtime behaviors for interfaces, which the language has to
+  support, and of course new interfaces and methods in core library that
+  leverages them. If you want to use generic math, you need a combination of
+  runtime, framework, and language that were designed to work in conjunction.
+
+* **library-level preview feature**. These are library APIs that don't require
+  any platform-level preview features to be consumed. The only difference to
+  regular APIs is that these are in preview so their shape isn't stable and thus
+  offers no backwards compatibility when upgrading to a later version of the
+  library.
+
+Due to their nature, platform-level preview features can only be consumed in a
+specific combination of framework and SDK: namely, the framework and SDK version
+must match. For example, you can only use preview features in `net6.0` when
+using the .NET 6 SDK and only use preview features in `net7.0` when using the
+.NET 7 SDK. Specifically, you can't use the .NET 7 SDK to use preview features
+in `net6.0`. That's because preview features aren't backwards compatible and
+platform-level preview features span both framework and tooling. In
+multi-targeting scenarios this means you can only use preview features for the
+latest framework version.
+
+For platform-level preview features this restriction makes sense because it
+models reality. However, for library-level preview features not only does this
+restriction not make sense, it's highly undesirable. Library authors simply want
+a way to express "these parts of my API surface might change". This statement
+isn't tied to any runtime-, framework-, or language features.
+
+Another concern we had for platform-level preview features is that we wanted to
+ensure that application developers are aware when their applications have a
+dependency on them. That is, we want to avoid a case where a library on NuGet
+uses platform-level preview features as an implementation detail that the
+application consumer isn't aware of. So we made the opt-in mechanism viral: a
+library that turns on platform-level preview features requires its consumer to
+turn them on as well.
+
+Since library-level preview features cannot change how the .NET runtime itself
+operates, we don't have that same level of concern. In fact, we believe the
+viral nature is also undesirable.
```

Comment:
A concern here from the Azure SDK side is that having it not be viral can lead to diamond dependency problems, where Library A exposes an experimental API, and then Libraries B and C both consume it, but at different versions where there are breaking changes.  Consumers who want to use both libraries B and C run into issues in that case.

Reply:
How does virality help with that? It would just meant that the consumer would also need to suppress `A.F1` and `B.F2`. There is logically nothing that would help with compilation or runtime errors around mismatched version of the shared diamond.

Reply:
The reason we don't think we need to make it viral goes like this:

1. If it bleeds into a consumer's public API (i.e. it's a type) then its consumer will need to suppress it anyway.

2. If it's not part of the API, then logically the code signs up for shielding its consumers from any breaking changes. If the code doesn't want that, they can apply the attribute to their entry points as well.

Reply:
It doesn't directly help once in that situation, but it helps a consumer of B and C to know that they are getting themselves into this danger zone, which they normally wouldn't expect with stable versioned dependencies.

This *could* be a guidance thing where we recommend that any API that suppresses the diagnostic for a call it makes, re-apply the attribute.  Alternatively, you could treat it like `Obsolete` where you have the choice to *either* suppress, or to make viral.

Reply:
I'm worried about how the error surfaces when using B and upgrading A to a version that changes or no longer supports the preview feature.  This will likely show up as missing method / type exceptions only at runtime.  How does the customer realize this is a problem with B and not A?  How can we get the customer to realize this is a problem at design time instead of runtime?

Reply:
@Pilchie @ericstj 

Thanks for providing more detail; I think you convinced me that virality is more desirable, because NuGet packages can be independently updated and there is no way for NuGet or the build system to provide actionable errors in case there is a mismatch in experimental API surface. I like @Pilchie suggestion of following `[Obsolete]` and `[RequiresRuntimeFeatures]` approach saying you either suppress (shield) or forward (mark yourself). That feels intuitive.

@geeknoid?

Reply:
How would viral behavior work then?

If my assembly calls any experimental APIs, then my assembly gets marked as an experimental assembly? Then when a customer tries to bind to my assembly, they get a warning about the fact my assembly uses experimental APIs in its implementation, and so the customer would then need to suppress that warning in order to allow the binding to happen. And then this would cause their assembly to be marked experimental as well.

Basically, we'd need ab assembly-level annotation to indicate "consumes experimental APIs", and then perhaps reflect this in the nuget package metadata as well somehow.

Is that about right?

Reply:
We could do better than a bool.  What if we had produces/consumes.  A produces experimental feature X.  B consumes experimental feature X.  If things line up everything is happy, no warnings.  If they get out of sync then an analyzer can flag B that it requires an experimental feature that's no longer present in A?

Reply:
That's basically how the original spec for named preview features worked. We could revive that.